File: internal/stt/silero.go
// Package stt provides speech-to-text functionality using sherpa-onnx.
// This file contains the Silero VAD implementation.
package stt

import (
	"fmt"
	"log"
	"path/filepath"
	"sync"
	"sync/atomic"
	"time"

	"github.com/agalue/sherpa-voice-assistant/internal/models"
	"github.com/agalue/sherpa-voice-assistant/internal/sherpa"
)

// VAD configuration constants.
const (
	// VADMinSpeechDuration is the minimum speech duration (in seconds) required to
	// trigger segment delivery. 0.1 s captures short utterances like "yes" or "no".
	VADMinSpeechDuration = 0.1

	// VADMaxSpeechDuration is the maximum continuous speech duration (in seconds).
	// Prevents unbounded audio accumulation and forces segmentation of long utterances.
	VADMaxSpeechDuration = 30.0

	// VADWindowSize is the VAD processing window in samples.
	// At 16 kHz: 512 samples = 32 ms (standard Silero VAD frame size).
	VADWindowSize = 512

	// VADBufferSize is the VAD audio buffer depth in seconds.
	VADBufferSize = 60.0
)

// Compile-time interface compliance check.
var _ VoiceDetector = (*SileroVAD)(nil)

// SileroVAD implements [VoiceDetector] using the Silero VAD model via sherpa-onnx.
//
// AcceptWaveform is safe to call from the real-time audio callback thread; it never
// blocks. Completed speech segments are delivered on [SileroVAD.SegmentChannel].
type SileroVAD struct {
	vad        *sherpa.VoiceActivityDetector // VAD engine (fast, <10 ms per call)
	mu         sync.Mutex                    // Protects VAD access
	sampleRate int

	// Atomic speech-detection state — lock-free on the hot path.
	wasSpeaking atomic.Bool
	speechStart atomic.Int64 // Unix nanoseconds; 0 if not currently in speech

	// Event-driven segment delivery.
	segmentChan chan []float32
}

// SileroConfig holds configuration for [SileroVAD].
type SileroConfig struct {
	ModelDir        string  // Base model directory (silero_vad.onnx is resolved automatically)
	Threshold       float32 // VAD confidence threshold (0.0–1.0)
	SilenceDuration float32 // Silence duration in seconds before speech is considered ended
	SampleRate      int
	NumThreads      int
	Verbose         bool
}

// NewSileroVAD creates a [SileroVAD] that satisfies [VoiceDetector].
func NewSileroVAD(cfg *SileroConfig) (*SileroVAD, error) {
	modelPath := filepath.Join(cfg.ModelDir, "silero_vad.onnx")

	vadConfig := &sherpa.VadModelConfig{}
	vadConfig.SileroVad.Model = modelPath
	vadConfig.SileroVad.Threshold = cfg.Threshold
	vadConfig.SileroVad.MinSilenceDuration = cfg.SilenceDuration
	vadConfig.SileroVad.MinSpeechDuration = VADMinSpeechDuration
	vadConfig.SileroVad.MaxSpeechDuration = VADMaxSpeechDuration
	vadConfig.SileroVad.WindowSize = VADWindowSize
	vadConfig.SampleRate = cfg.SampleRate
	vadConfig.NumThreads = cfg.NumThreads
	vadConfig.Debug = 0
	if cfg.Verbose {
		vadConfig.Debug = 1
	}

	vad := sherpa.NewVoiceActivityDetector(vadConfig, VADBufferSize)
	if vad == nil {
		return nil, fmt.Errorf("failed to create Silero VAD")
	}

	return &SileroVAD{
		vad:         vad,
		sampleRate:  cfg.SampleRate,
		segmentChan: make(chan []float32, 5),
	}, nil
}

// AcceptWaveform feeds audio samples into the VAD and delivers completed speech
// segments immediately via [SileroVAD.SegmentChannel].
//
// Must never block; called on the real-time audio callback thread.
func (v *SileroVAD) AcceptWaveform(samples []float32) {
	v.mu.Lock()
	v.vad.AcceptWaveform(samples)
	isSpeech := v.vad.IsSpeech()

	// EVENT-DRIVEN: send completed segments without holding the VAD lock.
	if !v.vad.IsEmpty() {
		segment := v.vad.Front()
		v.vad.Pop()

		if len(segment.Samples) > 0 {
			samplesCopy := make([]float32, len(segment.Samples))
			copy(samplesCopy, segment.Samples)
			v.mu.Unlock()

			// Non-blocking send — never block the audio callback thread.
			select {
			case v.segmentChan <- samplesCopy:
			default:
				log.Println("⚠️ Segment channel full, dropping segment")
			}

			v.mu.Lock()
		}
	}
	v.mu.Unlock()

	// Speech-state tracking via atomics (lock-free on this side of the hot path).
	wasSpk := v.wasSpeaking.Load()
	if isSpeech && !wasSpk {
		log.Println("🎤 Speech started")
		v.speechStart.Store(time.Now().UnixNano())
		v.wasSpeaking.Store(true)
	} else if !isSpeech && wasSpk {
		if startNano := v.speechStart.Load(); startNano > 0 {
			duration := float64(time.Now().UnixNano()-startNano) / 1e9
			log.Printf("🎤 Speech ended (%.1fs)", duration)
		}
		v.wasSpeaking.Store(false)
	}
}

// SegmentChannel returns the channel on which completed speech segments are delivered.
func (v *SileroVAD) SegmentChannel() <-chan AudioSegment {
	return v.segmentChan
}

// IsSpeechDetected returns true when the VAD currently considers input to be active speech.
func (v *SileroVAD) IsSpeechDetected() bool {
	v.mu.Lock()
	defer v.mu.Unlock()
	return !v.vad.IsEmpty() || v.vad.IsSpeech()
}

// Clear resets the internal VAD state.
func (v *SileroVAD) Clear() {
	v.mu.Lock()
	defer v.mu.Unlock()
	v.vad.Clear()
}

// Close releases all resources held by the detector.
func (v *SileroVAD) Close() {
	v.mu.Lock()
	defer v.mu.Unlock()

	if v.segmentChan != nil {
		close(v.segmentChan)
		v.segmentChan = nil
	}
	if v.vad != nil {
		sherpa.DeleteVoiceActivityDetector(v.vad)
		v.vad = nil
	}
}

// ---------------------------------------------------------------------------
// SileroModelProvider
// ---------------------------------------------------------------------------

// SileroModelProvider implements [ModelProvider] for the Silero VAD model.
type SileroModelProvider struct{}

// Name returns the human-readable name of this VAD implementation.
func (p *SileroModelProvider) Name() string {
	return "Silero VAD"
}

// EnsureModels downloads silero_vad.onnx if it is absent from modelDir.
func (p *SileroModelProvider) EnsureModels(modelDir string, force bool) error {
	dest := filepath.Join(modelDir, "silero_vad.onnx")
	return models.DownloadFile(
		"https://github.com/k2-fsa/sherpa-onnx/releases/download/asr-models/silero_vad.onnx",
		dest, force,
	)
}

// VerifyModels returns a list of absent model file paths.
func (p *SileroModelProvider) VerifyModels(modelDir string) []string {
	path := filepath.Join(modelDir, "silero_vad.onnx")
	if !models.FileExists(path) {
		return []string{path}
	}
	return nil
}

File: rust-impl/src/stt/silero.rs
//! Silero VAD detector — voice activity detection.
//!
//! [`SileroVad`] implements [`super::VoiceDetector`] and detects speech in the audio
//! stream, delivering completed audio segments via an async channel. Pair with
//! [`crate::stt::WhisperRecognizer`] to form a complete STT pipeline.
//!
//! [`SileroModelProvider`] manages the single `silero_vad.onnx` model file.

use std::path::PathBuf;
use std::sync::Arc;
use std::sync::atomic::{AtomicBool, Ordering};
use std::time::Instant;

use anyhow::Result;
use parking_lot::Mutex;
// Alias to avoid collision with our public struct name.
use sherpa_rs::silero_vad::{SileroVad as SherpaVad, SileroVadConfig};
use tokio::sync::mpsc;
use tracing::{debug, info, warn};

use super::VoiceDetector;
use crate::config::AppConfig;

/// Minimum speech duration in seconds to emit a segment.
const MIN_SPEECH_DURATION: f32 = 0.1;

/// Maximum speech duration in seconds (prevents runaway segments).
const MAX_SPEECH_DURATION: f32 = 30.0;

/// VAD window size in samples (512 = 32 ms at 16 kHz).
const VAD_WINDOW_SIZE: i32 = 512;

/// Audio buffer depth in seconds.
const VAD_BUFFER_SIZE_SECONDS: f32 = 60.0;

/// VAD state; separated so the audio-callback lock covers only fast VAD operations.
struct VadState {
    vad: SherpaVad,
    was_speaking: bool,
    speech_start: Option<Instant>,
}

/// Silero VAD implementation of [`VoiceDetector`].
///
/// Feeds audio into the Silero VAD model and delivers completed speech segments
/// via an async channel. Designed for event-driven use — no polling required.
///
/// # Lock discipline
///
/// The inner mutex is held only during VAD calls (<10 ms). The channel send is
/// performed **without** holding the lock to avoid blocking the audio callback thread.
pub struct SileroVad {
    vad_state: Arc<Mutex<VadState>>,
    segment_tx: mpsc::Sender<Vec<f32>>,
    /// Flag that is `true` while the user is speaking (used for playback interruption).
    pub speaking: Arc<AtomicBool>,
}

impl SileroVad {
    /// Create a new Silero VAD detector.
    ///
    /// # Returns
    /// A `(SileroVad, segment_receiver)` tuple. Pass the receiver to
    /// [`crate::stt::spawn_transcription_task`].
    ///
    /// # Errors
    /// Returns an error if the VAD model cannot be initialised.
    pub fn new(config: &AppConfig) -> Result<(Self, mpsc::Receiver<Vec<f32>>)> {
        let provider = config.effective_stt_provider();

        let vad_model = config.model_dir.join("silero_vad.onnx");

        let vad_config = SileroVadConfig {
            model: vad_model.to_string_lossy().to_string(),
            threshold: config.vad_threshold,
            sample_rate: config.sample_rate,
            min_silence_duration: config.vad_silence_duration,
            min_speech_duration: MIN_SPEECH_DURATION,
            max_speech_duration: MAX_SPEECH_DURATION,
            window_size: VAD_WINDOW_SIZE,
            provider: Some(provider.as_sherpa_provider().to_string()),
            num_threads: Some(config.vad_threads.try_into().unwrap_or(1)),
            debug: config.verbose,
        };

        let vad = SherpaVad::new(vad_config, VAD_BUFFER_SIZE_SECONDS).map_err(|e| anyhow::anyhow!("Failed to initialize Silero VAD: {}", e))?;

        info!("VAD initialized successfully");

        let (segment_tx, segment_rx) = mpsc::channel(5);

        let detector = Self {
            vad_state: Arc::new(Mutex::new(VadState { vad, was_speaking: false, speech_start: None })),
            segment_tx,
            speaking: Arc::new(AtomicBool::new(false)),
        };

        Ok((detector, segment_rx))
    }

    /// Get a clone of the speaking flag for external use before wrapping in a trait object.
    pub fn speaking_flag(&self) -> Arc<AtomicBool> {
        self.speaking.clone()
    }
}

impl VoiceDetector for SileroVad {
    /// Feed raw PCM samples to the VAD.
    ///
    /// **Never blocks.** Completed segments are sent via channel after the VAD
    /// lock is released, keeping the audio callback thread free.
    fn accept_waveform(&self, samples: &[f32]) {
        let mut state = self.vad_state.lock();
        state.vad.accept_waveform(samples.to_vec());

        let is_speech = state.vad.is_speech();
        self.speaking.store(is_speech, Ordering::SeqCst);

        // Track speech transitions for logging.
        if is_speech && !state.was_speaking {
            state.speech_start = Some(Instant::now());
            info!("🎤 Speech started");
        } else if !is_speech
            && state.was_speaking
            && let Some(start) = state.speech_start.take()
        {
            info!("🎤 Speech ended ({:.1}s)", start.elapsed().as_secs_f32());
        }
        state.was_speaking = is_speech;

        // EVENT-DRIVEN: send completed segment without holding the VAD lock.
        if !state.vad.is_empty() {
            let segment = state.vad.front();
            state.vad.pop();

            if !segment.samples.is_empty() {
                debug!("Segment completed: {} samples", segment.samples.len());
                let samples_to_send = segment.samples.clone();
                drop(state); // Release lock before channel send.

                if let Err(e) = self.segment_tx.try_send(samples_to_send) {
                    warn!("Failed to send segment (channel full): {}", e);
                }
            }
        }
    }

    #[allow(dead_code)]
    fn speaking_flag(&self) -> Arc<AtomicBool> {
        self.speaking.clone()
    }
}

// ─────────────────────────────────────────────────────────────────────────────
// ModelProvider
// ─────────────────────────────────────────────────────────────────────────────

/// Manages the `silero_vad.onnx` model file.
pub struct SileroModelProvider;

impl super::ModelProvider for SileroModelProvider {
    fn name(&self) -> &'static str {
        "Silero VAD"
    }

    /// # Errors
    /// Returns an error if the download fails.
    fn ensure_models(&self, model_dir: &std::path::Path, force: bool) -> Result<()> {
        let vad_dest = model_dir.join("silero_vad.onnx");
        crate::setup::download_file("https://github.com/k2-fsa/sherpa-onnx/releases/download/asr-models/silero_vad.onnx", &vad_dest, force)
    }

    fn verify_models(&self, model_dir: &std::path::Path) -> Vec<PathBuf> {
        let vad_path = model_dir.join("silero_vad.onnx");
        if !vad_path.exists() { vec![vad_path] } else { vec![] }
    }
}
